Adopt more smart pointers in APISerializedScriptValue and VideoFrameCV

RupinMittal opened the pull request:
#### ac88fd005ae0e756a4f29b37e02666a839975b77
<pre>
Adopt more smart pointers in APISerializedScriptValue and VideoFrameCV
<a href="https://bugs.webkit.org/show_bug.cgi?id=288023">https://bugs.webkit.org/show_bug.cgi?id=288023</a>
<a href="https://rdar.apple.com/145174758">rdar://145174758</a>

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createFromPixelBuffer):
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Shared/API/APISerializedScriptValue.h:

Canonical link: <a href="https://commits.webkit.org/290823@main">https://commits.webkit.org/290823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eae234293bba82eaf74369b4fe35e2303b41699e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/124 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96098 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41859 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93139 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11041 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18952 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70001 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27531 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94090 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8408 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82542 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50327 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8179 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/107 "Found 2 new test failures: imported/w3c/web-platform-tests/cookies/schemeful-same-site/schemeful-websockets.sub.tentative.html imported/w3c/web-platform-tests/css/css-grid/grid-definition/grid-auto-fill-columns-001.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40989 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78501 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/124 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98077 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13416 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79011 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18555 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78210 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22720 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/79 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11505 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14404 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18297 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23629 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18023 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21484 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19802 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -503,18 +503,18 @@ static void copyPlane(uint8_t* destination, const uint8_t* source, uint64_t sour
 
     auto dataBaseAddress = pixelBuffer->bytes().data();
-    auto leakedBuffer = &pixelBuffer.leakRef();
+    RefPtr leakedBuffer = &pixelBuffer.leakRef();
```
It's still not perfect but I think this would look a little less weird than assigning the `leakRef()` output to a RefPtr.
1. Drop this variable here.

cdumez commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -503,18 +503,18 @@ static void copyPlane(uint8_t* destination, const uint8_t* source, uint64_t sour
 
     CVPixelBufferRef cvPixelBufferRaw = nullptr;
-    auto status = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, dataBaseAddress, width * 4, derefBuffer, leakedBuffer, nullptr, &cvPixelBufferRaw);
+    auto status = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, dataBaseAddress, width * 4, derefBuffer, leakedBuffer.get(), nullptr, &cvPixelBufferRaw);
```
2. Here pass `pixelBuffer.ptr()` instead of `leakedBuffer`

cdumez commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -503,18 +503,18 @@ static void copyPlane(uint8_t* destination, const uint8_t* source, uint64_t sour
     auto cvPixelBuffer = adoptCF(cvPixelBufferRaw);
     if (!cvPixelBuffer) {
-        derefBuffer(leakedBuffer, nullptr);
+        derefBuffer(leakedBuffer.get(), nullptr);
```
3. Drop this line.

cdumez commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -503,18 +503,18 @@ static void copyPlane(uint8_t* destination, const uint8_t* source, uint64_t sour
+        derefBuffer(leakedBuffer.get(), nullptr);
         return nullptr;
     }
     ASSERT_UNUSED(status, !status);
```
4. Before the assertion, call:
```
// The CVPixelBuffer now owns the pixelBuffer and will call `deref()` on it when it gets destroyed. 
pixelBuffer->ref();
```

This part is a bit ugly but I couldn't come up with a better idea yet and I think it is less confusing than assigning the `leakRef()` result to a `RefPtr`.

RupinMittal replied:
Done. Although I'm not entirely sure what's happening here. Could you explain why we need
`pixelBuffer->ref();` here?

RupinMittal committed:
Adopt more smart pointers in APISerializedScriptValue and VideoFrameCV
https://bugs.webkit.org/show_bug.cgi?id=288023
rdar://145174758

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createFromPixelBuffer):
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Shared/API/APISerializedScriptValue.h:

Canonical link: https://commits.webkit.org/290823@main
